### DISCUSSION TITLE
Force Page Reload After Table Button Action

### DISCUSSION THREAD

**Discussion by intrepidws** [filamentphp/filament#1694] [Help] [answered]
Is there a way to force a page reload after a button action on a table? I'm using the button action to set a session variable, so I need the page to reload after it's done.

My code:

```
ButtonAction::make('add_to_session')
    ->label('Make This Active')
    ->action(function ($record) {
        session()->put('active_id', $record->id);
        
        $this->notify('success', 'This has been set to active.');
    })
```

**Comment by studiowizjo** [comment]
Put this in your action callback:

```php
$this->getTableRecords()->fresh();
```

**Reply by intrepidws** [reply]
Unfortunately, this didn't work for me. It might be specific to my use case, since I'm setting a session variable and the refresh needs to be aware of that new session variable. Thanks anyway.

**Comment by intrepidws** [comment, marked as answer]
I ended up creating a route to handle my action, and then redirecting back. In case it's helpful to anyone, this was the button action code:

```php
TableButtonAction::make('print')
    ->label('Make Printer Active')
    ->url(function (Printer $record) {
        return route('printer.activate', [$record]);
    })
    ->visible(function (Printer $record) {
        return $record->isNotSessionScoped();
    });
```

And then within the controller action, I was able to still set the Filament Flash notification using this code:

```php
Filament::notify('success', 'Printer has been set to active.', isAfterRedirect: true);
```

**Comment by studiowizjo** [comment]
If you need to not only refresh the data, but also alter the query, you can use instead this method:

```php
$this->getTableQuery() 
```

and chain to it your conditions.

**Reply by intrepidws** [reply]
I did some more testing with the snippet of `$this->getTableRecords()->fresh();` that you recommended - and it _does_ work. Well, it does exactly what you describe - but it doesn't really work for my purposes. I am fairly certain that the issue on my end is that I'm setting a session variable and the reason that I need the refresh is to change the page based on the fact that the session variable is now set. If I click the button once, the refresh page isn't aware of the newly set session variable - but if I press the button twice, it is. So I'm assuming it's an issue of session timing, etc.

Thanks again for your help here @studiowizjo

**Reply by studiowizjo** [reply]
From what you are saying your db query is based on something that is being saved in the session, so before the fresh data can be retrieved based on it, you can modify your query with an additional condition - the same you write to the session.

The 1st method I described assumes, that query doesnt change and it only gets fresh data from db, on the same conditions which where while loading the page:

```php
$this->getTableRecords()->fresh();
```

but in your case the 2nd method should be solution, because this way you can modify your query, the same way you will do while loading page:

```php
$this->getTableQuery()->where('active_id', $record->id) //or whatever condition you need to apply.
```

**Reply by intrepidws** [reply]
Ohhhhh, now I get it. Sorry, a bit slow on the uptake here - that's very helpful, thank you.
